src/back/spv/writer.rs

```diff
@@ -1075,199 +1160,269 @@ impl Writer {
+                            id: variable_id,
+                            name: None,
+                            instruction: self.instruction_variable(
+                                pointer_id,
+                                variable_id,
+                                spirv::StorageClass::Function,
+                                None,
+                            ),
+                        });
+                        block.body.push(self.instruction_store(variable_id, id));
+                        argument_ids.push(variable_id);
+                    }
+
+                    let return_type_id = self
+                        .get_function_return_type(origin_function.return_type, &ir_module.types);
+
+                    block.body.push(self.instruction_function_call(
+                        return_type_id,
+                        id,
+                        *self.lookup_function.get(local_function).unwrap(),
+                        argument_ids,
+                    ));
+                    Some((id, None))
+                }
+                _ => unimplemented!("{:?}", origin),
+            },
             _ => unimplemented!("{:?}", expression),
         }
     }
 
-    fn write_function_block(
+    fn write_function_statement(
         &mut self,
         ir_module: &crate::Module,
-        function: &crate::Function,
+        ir_function: &crate::Function,
         statement: &crate::Statement,
-        output: &mut Vec<Instruction>,
-    ) -> Instruction {
+        block: &mut Block,
+        function: &mut Function,
+    ) {
         match statement {
-            crate::Statement::Return { .. } => match function.return_type {
-                Some(ty) => {
-                    let value_id = self.get_type_id(&ir_module.types, LookupType::Handle(ty));
-                    self.instruction_return_value(value_id)
+            crate::Statement::Return { value } => match ir_function.return_type {
+                Some(_) => {
+                    let expression = &ir_function.expressions[value.unwrap()];
+                    let (id, ty) = self
+                        .write_expression(ir_module, ir_function, expression, block, function)
+                        .unwrap();
+
+                    let id = match expression {
+                        crate::Expression::LocalVariable(_) => {
+                            let load_id = self.generate_id();
+                            let value_ty_id = self
+                                .get_type_id(&ir_module.types, LookupType::Handle(ty.unwrap()));
+                            block.body.push(self.instruction_load(
+                                value_ty_id,
+                                load_id,
+                                id,
+                                None,
+                            ));
+                            load_id
+                        }
+                        _ => id
+                    };
```

Comment:
The only thing I am not yet happy about is this piece of code. This is also used (for now) at `Statement::Store`. The reason for manually checking is because if the following conclusion I made:

There is a difference between what id to pass, a local variable is in an unique case I think that it can be both at the left side and the right side of an operator, e.g:

```wgsl
var test: f32 = 1.0; // Left side of the operator
var test_2: f32 = test; // Right side of the operator
```

So the reason for an OpLoad here is because if a variable is used at the right side of an operator (an expression), then we need to load it. If it is referenced/used as an actual variable reference, then we need to have its direct id.

Above example outputs the following spirv:
```spirv
%4 = OpVariable %_ptr_Function_float Function %float_1
%6 = OpVariable %_ptr_Function_float Function
%11 = OpLoad %float %4 None
OpStore %6 %11
```

I think this all sounds pretty logical, and I believe this case only exists for variables, but I wanted to give attention to this piece of code.

Reply:
that makes sense! I'm not sure if we'll need to special case any global variable types as well, but what you have sounds reasonable